Reuse registration on network error

chmielewskiandreas opened the pull request:
Based on PR https://github.com/zephyrproject-rtos/zephyr/pull/45476/commits/b662799fe96706bed8f0c6f1db23d0c7a5837815 this PR covers the situation to reuse a registration on a network error. In a later patch my goal is to offer the application more control over network errors and to be able to suspend/resume connection in non-queue-mode

andreas-chmielewski-gcx committed:
net: lwm2m: Update registration only in good path

Registration should only be updated if update of the registration was
succesful.

Signed-off-by: Andreas Chmielewski <andreas.chmielewski@grandcentrix.net>

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -258,15 +268,19 @@ static void socket_fault_cb(int error)
 	 * bootstrap procedure.
 	 */
-	set_sm_state(ENGINE_DO_REGISTRATION);
+	set_sm_state(ENGINE_NETWORK_ERROR);
```
Comment above is now misleading. Just remove it.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -318,6 +332,8 @@ static int do_bootstrap_reply_cb(const struct coap_packet *response,
 
 	if (code == COAP_RESPONSE_CODE_CHANGED) {
+		/* remember the last reg time */
+		client.last_update = k_uptime_get();
```
I don't think we should update the timer here. Bootstrapped is not yet registered.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -1065,7 +1086,23 @@ static void sm_do_network_error(void)
+	}
+
+	engine_trigger_update(false);
+	set_sm_state(ENGINE_REGISTRATION_DONE);
```
This state change worries me.
`NETWORK_ERROR ---(invalid state change)---> REGISTRATION_DONE --> UPDATE_SENT`

I see that by doing this, you are triggering the functionality in `sm_registration_done()`. But this is now hard to spot, and not obvious at all for developer looking at this code. If we need calls from `sm_registration_done()` maybe it needs to split out from there, so we can directly call the function and stay correctly in `NETWORK_ERROR` state. Or maybe we should consider going into `RX_OFF` or something like that.

SeppoTakalo reviewed (changes requested):
I would like to see the error handling done in error state without causing any side effects of transitioning the client into REGISTERED state when in fact, it is not registered.

andreas-chmielewski-gcx committed:
net: lwm2m: Try to reuse registration in case of a network error

For most of the times it is not necessary to do a full registration
once a connection is established after a network error. This is in
particular not needed if lifetime is not yet expired and the server
does not refuse a registration update.

Signed-off-by: Andreas Chmielewski <andreas.chmielewski@grandcentrix.net>

chmielewskiandreas commented:
Any update/comments on my last commit?

rlubos reviewed (commented):
Looks fine to me, but let's wait for @SeppoTakalo feedback as he was mostly involved in the review.

chmielewskiandreas commented:
@SeppoTakalo What do you think about the changes I did?

chmielewskiandreas commented:
> @SeppoTakalo What do you think about the changes I did?

I would like to hear your opinion before I do any kind of rebase here and invest time for that. Thanks :)

SeppoTakalo reviewed (dismissed):
I think this now looks as intended.

Can you also update the documentation and a diagram from here:
https://docs.zephyrproject.org/latest/connectivity/networking/api/lwm2m.html#lwm2m-engine-and-application-events

The diagram is in a .png format that can be opened in [draw.io](https://app.diagrams.net/) then add the add the new "UPDATE_REGISTRATION" to the table as well.

andreas-chmielewski-gcx committed:
net: lwm2m: separate closing context from closing socket

Separate closing lwm2m context from closing socket. This patch is required
for the rd client to take more control over lwm2m context and the socket.
The goal is to close the socket and to keep the lwm2m context if this is
needed.

Signed-off-by: Andreas Chmielewski <andreas.chmielewski@grandcentrix.net>

chmielewskiandreas commented:
> I think this now looks as intended.
> 
> Can you also update the documentation and a diagram from here: https://docs.zephyrproject.org/latest/connectivity/networking/api/lwm2m.html#lwm2m-engine-and-application-events
> 
> The diagram is in a .png format that can be opened in [draw.io](https://app.diagrams.net/) then add the add the new "UPDATE_REGISTRATION" to the table as well.

Thx :) Due to refactoring I had to do minor refactoring (patch net: lwm2m: separate closing context from closing socket). I hope this approach is okay for you as well.

SeppoTakalo commented:
@jheiskan81 can you review this also. I'm wondering if this clashes the same areas as your work on pause-resume topic.

jheiskan81 reviewed (dismissed):
Yes true. https://github.com/zephyrproject-rtos/zephyr/pull/47327 Pause and stop functionality have a similar functionality:
Stop clear context and reset observation.
Pause only suspend connection and suspend thread.
Resume function validate similar way lifetime and enable thread: If still valid jump to update process otherwise to Full registration. There will be minor conflict but not a huge.

SeppoTakalo commented:
build failures:
```
zephyr/subsys/net/lib/lwm2m/lwm2m_rd_client.c:190:3: error: implicit declaration of function 'lwm2m_socket_close' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                lwm2m_socket_close(client.ctx);
               
```

chmielewskiandreas commented:
> build failures:
> 
> ```
> zephyr/subsys/net/lib/lwm2m/lwm2m_rd_client.c:190:3: error: implicit declaration of function 'lwm2m_socket_close' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
>                 lwm2m_socket_close(client.ctx);
>                
> ```

fixed :)

SeppoTakalo reviewed (approved):
Thank you for the thorough work you have done with this.
This is now looking clean.

chmielewskiandreas commented:
> Thank you for the thorough work you have done with this. This is now looking clean.

thanks a lot for your reviews :)

rlubos reviewed (approved):
LGTM, thanks.
